pkg/annotations/middleware.go

```diff
@@ -7,55 +7,51 @@ import (
 	"github.com/containous/traefik/v2/pkg/config/dynamic"
 )
 
-type middlewareBuilder func(middleware *dynamic.Middleware, annotations map[string]string) (bool, error)
+type middlewareBuilder func(annotations map[string]string) (*dynamic.Middleware, string, error)
+
+// BuildMiddlewares builds middlewares from the given annotations.
+func BuildMiddlewares(annotations map[string]string) (map[string]*dynamic.Middleware, error) {
 
-// BuildMiddleware builds a middleware from the given annotations
-func BuildMiddleware(annotations map[string]string) (*dynamic.Middleware, error) {
 	builders := []middlewareBuilder{
 		buildRetryMiddleware,
 		buildRateLimitMiddleware,
 		buildCircuitBreakerMiddleware,
 	}
 
-	var (
-		middleware        dynamic.Middleware
-		middlewareCounter int
-	)
+	middlewares := map[string]*dynamic.Middleware{}
 
 	for _, builder := range builders {
-		ok, err := builder(&middleware, annotations)
+		middleware, name, err := builder(annotations)
 		if err != nil {
 			return nil, err
 		}
 
-		if ok {
-			middlewareCounter++
+		if middleware != nil {
+			middlewares[name] = middleware
 		}
 	}
 
-	if middlewareCounter == 0 {
-		return nil, nil
-	}
-
-	return &middleware, nil
+	return middlewares, nil
 }
 
-func buildRetryMiddleware(middleware *dynamic.Middleware, annotations map[string]string) (bool, error) {
+func buildRetryMiddleware(annotations map[string]string) (*dynamic.Middleware, string, error) {
```

Comment:
Please use named returns (but not naked returns) to make it clearer what this function will return from reading its prototype